pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,282 @@
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/commonarea"
+	"antrea.io/antrea/pkg/antctl/raw/multicluster/common"
+	multiclusterscheme "antrea.io/antrea/pkg/antctl/raw/multicluster/scheme"
+)
+
+type clusterOptions struct {
+	Namespace  string `yaml:"namespace,omitempty"`
+	ClusterID  string `yaml:"clusterID,omitempty"`
+	Secret     string `yaml:"secret,omitempty"`
+	Server     string `yaml:"server,omitempty"`
+	ClusterSet string `yaml:"clusterSet,omitempty"`
+}
+
+type joinOptions struct {
+	LeaderClusterOpt clusterOptions `yaml:"leaderCluster"`
+	MemberClusterOpt clusterOptions `yaml:"memberCluster"`
+	ConfigFile       string
+}
+
+type meta struct {
+	Kind       string `yaml:"kind"`
+	ApiVersion string `yaml:"apiVersion"`
+}
+
+var joinOpts *joinOptions
+
+func (o *joinOptions) validateAndComplete() error {
+	if o.ConfigFile != "" {
+		raw, err := ioutil.ReadFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		typeMeta := &meta{}
+		if err := yamlUnmarshall(raw, typeMeta); err != nil {
+			return err
+		}
+		if typeMeta.Kind != "ClusterSetConfig" || typeMeta.ApiVersion != "multicluster.antrea.io/v1alpha1" {
+			return fmt.Errorf("unknown apiVersion or kind in config: %s", o.ConfigFile)
+		}
+		if err := yamlUnmarshall(raw, &joinOpts); err != nil {
+			return err
+		}
+	}
+
+	if o.LeaderClusterOpt.ClusterID == "" {
+		return fmt.Errorf("the leader clusterID cannot be empty")
+	}
+	if o.LeaderClusterOpt.Server == "" {
+		return fmt.Errorf("the leader server cannot be empty")
+	}
+	if o.LeaderClusterOpt.Secret == "" {
+		return fmt.Errorf("the leader secret cannot be empty")
+	}
+	if o.LeaderClusterOpt.Namespace == "" {
+		return fmt.Errorf("the leader namespace cannot be empty")
+	}
+	if o.MemberClusterOpt.ClusterSet == "" {
+		return fmt.Errorf("the member cluster set cannot be empty")
+	}
+	if o.MemberClusterOpt.ClusterID == "" {
+		return fmt.Errorf("the member cluster id cannot be empty")
+	}
+	if o.MemberClusterOpt.Namespace == "" {
+		return fmt.Errorf("the member namespace cannot be empty")
+	}
+
+	return nil
+}
+
+func yamlUnmarshall(raw []byte, v interface{}) error {
+	d := yaml.NewDecoder(bytes.NewReader(raw))
+	d.KnownFields(false)
+	return d.Decode(v)
+}
+
+var joinExamples = strings.Trim(`
+# Join the ClusterSet from a member cluster with command line options.
+  $ antctl mc join --leader-namespace=<LEADER_NAMESPACE> \
+				--leader-cluster-id=<LEADER_CLUSTER_ID> \
+				--secret=<SECRET> \
+				--leader-server=<LEADER_SERVER> \
```

Comment:
At least for this one we better to provide a real example like "https://172.18.0.3:6443". Would you change all commands to use real values? Actually that way is consistent with other antctl commands.

Reply:
Sure, I will change the example to real argument later. Thanks

Reply:
changed all the examples to real arguments.